Store some FIPS global variables in the FIPS_GLOBAL structure

mattcaswell opened the pull request:
We had some FIPS global variables that were based on values from the
config file. In theory if two instances of the fips module are loaded
they could be based on different config files which would cause this to
fail. Instead we store them in the FIPS_GLOBAL structure.

Fixes #14364

FdaSilvaYY commented on providers/fips/fipsprov.c:
```diff
@@ -634,10 +643,34 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
     fgbl->handle = handle;
 
+    /*
+     * We did initial set up of selttest_params in a local copy, because we
```
typo: selttest_params

slontis commented on providers/common/securitycheck_default.c:
```diff
@@ -17,13 +17,13 @@
 
 /* Disable the security checks in the default provider */
-int ossl_securitycheck_enabled(void)
+int ossl_securitycheck_enabled(OSSL_LIB_CTX *libctx)
```
ossl_unused?

mattcaswell replied:
Not sure why we need to do that here? We have lots of places where we have unused parameters and don't mark them.

mattcaswell replied:
> It's good practice to?

Is it? Why?

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -77,12 +72,18 @@ static OSSL_FUNC_core_get_libctx_fn *c_get_libctx = NULL;
 
 static void *fips_prov_ossl_ctx_new(OSSL_LIB_CTX *libctx)
 {
     FIPS_GLOBAL *fgbl = OPENSSL_zalloc(sizeof(*fgbl));
```
do null check here

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -634,10 +643,34 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+        /* Error already raised */
+        return 0;
+    }
+    /* Disable the conditional error check if is disabled in the fips config file*/
```
may as well fix up the */

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -634,10 +643,34 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+        && strcmp(fgbl->selftest_params.conditional_error_check, "0") == 0)
+        SELF_TEST_disable_conditional_error_state();
+
+    /* Disable the security check if is disabled in the fips config file*/
```
*/ here also..

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -819,16 +852,18 @@ int BIO_snprintf(char *buf, size_t n, const char *format, ...)
+    FIPS_GLOBAL *fgbl = ossl_lib_ctx_get_data(libctx,
+                                              OSSL_LIB_CTX_FIPS_PROV_INDEX,
+                                              &fips_prov_ossl_ctx_method);
+    return fgbl->fips_security_checks;
```
blankline

slontis commented:
ansi failure is relevant..

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -37,19 +38,13 @@ static OSSL_FUNC_provider_query_operation_fn fips_query;
- * ever be different from one init to the next? Unfortunately we can't do this
- * at the moment because c_put_error/c_add_error_vdata do not provide
- * us with the OSSL_LIB_CTX as a parameter.
+ * Should these function pointers be stored in the provider side provctx? Could
```
This is the reason I didn't try doing this PR myself :) Passing the libctx to these would be painful.

t8m replied:
When setting the function pointers, should we check the values and refuse loading the provider if the current value is non-NULL and different from the new value?

mattcaswell replied:
I've added a sanity check to ensure the up-call function pointers haven't changed.

paulidale replied:
I think that's good enough for 3.0.

t8m commented on providers/common/provider_seeding.c:
```diff
@@ -18,26 +18,30 @@ static OSSL_FUNC_cleanup_nonce_fn *c_cleanup_nonce = NULL;
+        /*
+        * We do not support the scenario of an application linked against
+        * multiple versions of libcrypto (e.g. one static and one dynamic), but
+        * sharing a single fips.so. We do a simple sanity check here.
```
This is common for all providers so perhaps `sharing a single provider.so.`

t8m replied:
Also the indentation is off by one space.

mattcaswell replied:
Actually it only impacts the fips provider. The built-in providers (default, base, null) will always be associated with the version of libcryto they are built into. So it should never impact those. The legacy provider is linked against libcrypto and doesn't use any up-calls, so it doesn't impact that one either. That only leaves fips.so as being affected by this.

t8m replied:
Ah, ok, so only the indentation... :grin:

t8m commented on providers/fips/fipsprov.c:
```diff
@@ -542,88 +542,103 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+        * multiple versions of libcrypto (e.g. one static and one dynamic), but
+        * sharing a single fips.so. We do a simple sanity check here.
+        */
+#define set_func(c, f) if (c != NULL && c != f) goto err; else c = f;
```
`return 0;` instead of `goto err;` otherwise you would have to do `*provctx = NULL` but that makes no sense as the err block does not do anything useful here.

So I'd write it as: `#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;`

t8m commented on providers/common/provider_seeding.c:
```diff
@@ -18,26 +18,30 @@ static OSSL_FUNC_cleanup_nonce_fn *c_cleanup_nonce = NULL;
+        * multiple versions of libcrypto (e.g. one static and one dynamic), but
+        * sharing a single fips.so. We do a simple sanity check here.
+        */
+#define set_func(c, f) if (c != NULL && c != f) goto err; else c = f;
```
Tiny little bit more efficient way is:
`#define set_func(c, f) if (c == NULL) c = f; else if (c != f) goto err;`

t8m commented on providers/legacyprov.c:
```diff
@@ -179,28 +175,12 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
-    if (c_get_libctx == NULL)
-        return 0;
+    /*
+     * We do not need to use an up-calls proviced by libcrypto, so we ignore
```
`to use up-calls provided`

mattcaswell commented:
All feedback addressed. Please see latest fixup.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -506,114 +535,117 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+         * multiple versions of libcrypto (e.g. one static and one dynamic), but
+         * sharing a single fips.so. We do a simple sanity check here.
+         */
+#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;
```
Is it possible to raise an error here if the error callback exists?  Presumably, the callback will exist since this code has been done previously.

Or is it not worth the effort?

mattcaswell replied:
Same as https://github.com/openssl/openssl/pull/14814#discussion_r612202109

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -634,10 +666,37 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+        return 0;
+    }
+    /*
+     * Disable the conditional error check if is disabled in the fips config
```
@slontis, why are we keeping these?

slontis replied:
What do you mean?
There are are few configurable items that were requested. This was one of them.

paulidale replied:
Turning them off kind of invalidates FIPS.

slontis replied:
Yep.
Needs to be added to security policy.

paulidale replied:
We should make a list of "things to go in the security policy"

mattcaswell commented on providers/common/provider_seeding.c:
```diff
@@ -18,22 +18,24 @@ static OSSL_FUNC_cleanup_nonce_fn *c_cleanup_nonce = NULL;
+         * multiple versions of libcrypto (e.g. one static and one dynamic), but
+         * sharing a single fips.so. We do a simple sanity check here.
+         */
+#define set_func(c, f) if (c == NULL) c = f; else if (c != f) return 0;
```
It's possible - although the error will go on the stack of the "other" libcrypto - so if an application calls "OSSL_PROVIDER_load" and encounters this problem then examining the error stack will not show any errors.

paulidale replied:
Yeah, good point.  Ideally, the error could go to both but that too hard.

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented on providers/common/securitycheck_default.c:
```diff
@@ -17,13 +17,13 @@
-int ossl_digest_rsa_sign_get_md_nid(const EVP_MD *md,
-                                    ossl_unused int sha1_allowed)
+int ossl_digest_rsa_sign_get_md_nid(OSSL_LIB_CTX *ctx, const EVP_MD *md,
+                                    int sha1_allowed)
```
Removed in error. Put back.

mattcaswell commented:
Rebased to resolve the conflict with master. Also added back the `ossl_unused` modifier that was removed in error. However I've not added any new ones (I'd rather not litter our code unnecessarily with these things).

@t8m please can you reconfirm?

t8m reviewed (approved):
Still approved

mattcaswell commented:
Pushed. Thanks.
